lib/logthrsource/logthrsourcedrv.c

```diff
@@ -249,39 +253,61 @@ log_threaded_source_driver_init_method(LogPipe *s)
   LogThreadedSourceDriver *self = (LogThreadedSourceDriver *) s;
   GlobalConfig *cfg = log_pipe_get_config(s);
 
-  self->worker = log_threaded_source_worker_new(cfg);
+  self->workers = g_new0(LogThreadedSourceWorker *, self->num_workers);
+  for (size_t i = 0; i < self->num_workers; i++)
+    self->workers[i] = self->worker_construct(self, i);
 
   if (!log_src_driver_init_method(s))
-    return FALSE;
+    goto error;
 
   g_assert(self->format_stats_key);
 
-  StatsClusterKeyBuilder *kb = stats_cluster_key_builder_new();
-  self->format_stats_key(self, kb);
   log_threaded_source_worker_options_init(&self->worker_options, cfg, self->super.super.group);
-  log_threaded_source_worker_set_options(self->worker, self, &self->worker_options,
-                                         self->super.super.id, kb);
 
-  LogPipe *worker_pipe = log_threaded_source_worker_logpipe(self->worker);
-  log_pipe_append(worker_pipe, s);
-  if (!log_pipe_init(worker_pipe))
+  for (size_t i = 0; i < self->num_workers; i++)
     {
-      log_pipe_unref(worker_pipe);
-      self->worker = NULL;
-      return FALSE;
+      StatsClusterKeyBuilder *kb = stats_cluster_key_builder_new();
+      self->format_stats_key(self, kb);
+      _worker_set_options(self->workers[i], self,
+                          &self->worker_options, self->super.super.id, kb);
+
+      LogPipe *worker_pipe = _worker_logpipe(self->workers[i]);
+      log_pipe_append(worker_pipe, s);
+      if (!log_pipe_init(worker_pipe))
+        goto error;
     }
 
   return TRUE;
+
+error:
+  for (size_t i = 0; i < self->num_workers; i++)
+    {
+      LogPipe *worker_pipe = _worker_logpipe(self->workers[i]);
+      if (!worker_pipe)
+        break;
+
+      log_pipe_deinit(worker_pipe);
+      log_pipe_unref(worker_pipe);
+      self->workers[i] = NULL;
+    }
```

Comment:
maybe it would be better to extract this into a separate function and combine this loop here and the one in deinit().

There's also a potential memory leak: the workers array is not freed here, which probably should be.

Reply:
Good idea, nice catch!

I have factored out some worker related logic to `_create_workers()`, `_init_workers()` and `_destroy_workers()`.